### ISSUE TITLE
Support for slice assign

### ISSUE THREAD

**Issue by louisfd** [huggingface/candle#1351]
I'm bringing back to life issue #663 
The proposed solution of slice-scatter in #927 does not seem to be enough. 
I've tried turning things around but I don't succeed in accomplishing the following: 

```rust
let a =  Tensor::new([[0.0, 1.0, 2.0], [3.0, 4.0, 5.0]]);
let b = Tensor::new( [[88.0, 99.0]]);
let new_a = a.slice_assign(b, [1..2, 1..3])
assert_eq!(new_a, [[0.0, 1.0, 2.0], [3.0, 88.0, 99.0]])
```

I think I'm not able to achieve this with slice_scatter because it only takes into account one dimension at a time.

**Comment by EricLBuehler** [comment]
Just saw this, but relates to #1279.

**Comment by LaurentMazare** [comment]
> ```rust
> let a =  Tensor::new([[0.0, 1.0, 2.0], [3.0, 4.0, 5.0]]);
> let b = Tensor::new( [[88.0, 99.0]]);
> let new_a = a.slice_assign(b, [1..2, 1..3])
> assert_eq!(a, [[0.0, 1.0, 2.0], [3.0, 88.0, 99.0]])
> ```

Thanks for bringing this up. Just to be sure, should the `assert_eq` in the example be about `new_a` or about `a`? Tensors are immutable by design in candle so it's unlikely that we will get the snippet you gave to work on `a` (we have some thoughts about providing in-place operations for variables rather than tensors but would really prefer avoiding it).
If it's about `new_a` it's probably easier to have but it would be good to have some example of models using this to see if it isn't easy to get some workaround there? (it seems like a bit of a convoluted operation so may be easier if we can avoid it and use simpler ops instead but that really depends on what you're trying to achieve here)

**Comment by EricLBuehler** [comment]
> In Burn we use slice_assign for instance in [autoregressive attention](https://github.com/Tracel-AI/burn/blob/main/burn-core/src/nn/attention/mask.rs) and [LSTM](https://github.com/Tracel-AI/burn/blob/main/burn-core/src/nn/rnn/lstm.rs). 

candle-sampling needs this slice_assign feature for the implementation of Beam search.

**Comment by LaurentMazare** [comment]
> Hi Laurent, Of course I meant `new_a`, sorry for the typo. I fixed it now. In Burn we use slice_assign for instance in [autoregressive attention](https://github.com/Tracel-AI/burn/blob/main/burn-core/src/nn/attention/mask.rs) and [LSTM](https://github.com/Tracel-AI/burn/blob/main/burn-core/src/nn/rnn/lstm.rs). But even more important is that it's in the backward pass of our equivalent of Candle's `narrow` operator.

Ok, I've merged #1377 that adds some pretty naive implementation for `slice_assign`, it's quite inefficient but avoids introducing some "base op". If this ends up being a bottleneck for performance, we could consider making something more efficient (but I kind of doubt that this will end up being in the performance critical op).

**Comment by LaurentMazare** [comment]
Closing again as hopefully all good now :)
